src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs

```diff
@@ -1179,6 +1179,15 @@ public StringBuilder Append(char value)
                 return this;
             }
 
+            unsafe
+            {
+                Span<char> tempBuffer = stackalloc char[32]; // should be enough for all primitives
+                if (value.TryFormat(tempBuffer, out int charsWrittenToTmp, format: default, provider: null))
+                {
+                    return Append(tempBuffer.Slice(0, charsWrittenToTmp));
+                }
+            }
+
```

Comment:
Debug.Fail?

Reply:
I was thinking of `Debug.Fail` but is it actually a failure? It probably makes sense to have a CI run with it and remove after

Reply:
If it's not a failure then the comment above is wrong.

Reply:
I just saw your comment in the PR description: "I'm leaving the string fallback for marginal cases".  If it's possible the size is could be too small, which it sounds like it is (e.g. you could have your current culture set to have a 40 character-long negative sign :), then the comment should instead be updated to say that it's good enough for most calls.

(I'm still not entirely convinced this extra code path is worthwhile, as if we get here we're about to allocate a new StringBuilder, a new char[], etc., and it should only happen relatively rarely... but if you've done the measurements to prove otherwise, ok.)

Reply:
@stephentoub 
```csharp
    public StringBuilder Test1()
    {
        var sb = new StringBuilder();
        for (int i = 100; i < 10000; i++)
            sb.Append(i);
        return sb;
    }
```
```
Allocated: 79.89 KB
Allocated: 79.48 KB
```
at a cost of minor perf regression since `stackalloc` (and additional code in general) makes `AppendSpanFormattable` non inlineable.

Reply:
Code that cares about allocations with StringBuilder typically pools them, and with a pooled builder, you're going to hit this case even more rarely because it will generally have grown to accomodate the kind of state being appended.  But even without that, your example shows a 0.5% memory saving (and it doesn't factor in the fact that if you create a StringBuilder, you're also about to call ToString on it).  Given that, it still seems like the right tradeoff here is to keep the code simple and optimize for throughput rather than the last mile of allocation.  Different opinions?

Reply:
I have no opinions on this, I've fixed the overhead by moving the whole "slow" path into a nested function.